### DISCUSSION TITLE
Feature request: ToAsciiString() - returns a string without colors and special characters

### DISCUSSION THREAD

**Discussion by Sire** [spectreconsole/spectre.console#1125] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
I want to render a colorful ANSI table to the console AND also save the table as ASCII to a textfile for later.

When looking at the source, I cannot find an easy way to solve this, and searching API documentation, StackOverflow and issues here turns up nothing. Perhaps there is a simple solution I've missed?

**Describe the solution you'd like**
Optimally, there would be a method on supported widgets that forced a non-ansi, no-color output, like:
```
var table = CreateTable();
AnsiConsole.Write(table);
string ascii = table.ToAsciiString();
```
This would create a custom RenderOptions with ColorSystem = ColorSystem.NoColors and Ansi=false, a default width (overridable) and render the simplest ASCII table.

**Describe alternatives you've considered**

Redirecting console output works but feels like overkill, and you cannot use markdown codes (anything that creates ANSI). 

```
FileStream fs = new FileStream("Output.txt", FileMode.Create);
TextWriter tmp = Console.Out;
StreamWriter sw = new StreamWriter(fs);
Console.SetOut(sw);
Console.WriteLine("This will be saved to file, but you cannot use ANSI codes");
AnsiConsole.Write(CreateTable());
```

Thanks for a great product!

**Comment by patriksvensson** [comment, marked as answer]
This can be done by using a `Recorder` which was designed for this purpose.

### Using the static API

```csharp
// Start recording
AnsiConsole.Record();

// Write stuff to the console
AnsiConsole.Write(new Table() ... );

// Get output as string
string text = AnsiConsole.ExportText();

// Export using a custom encoder
string custom = AnsiConsole.Export(new MyOutputEncoder());
```

### Using `IAnsiConsole`

```csharp
using(var recorder = new Recorder(console)) 
{
	// Write stuff to the console
	console.Write(new Table() ... );

	// Get output as string
	string text = recorder.ExportText();

	// Export using a custom encoder
	string custom = recorder.Export(new MyOutputEncoder());
}
```

**Reply by patriksvensson** [reply]
Update: We have a [ToAnsi](https://github.com/spectreconsole/spectre.console/blob/main/src/Spectre.Console/Extensions/Advanced/AnsiConsoleExtensions.cs#L32) extension method that takes an IRenderable. I guess we could add a `ToAscii` method as well here.

**Reply by TLevitt** [reply]
Hi Patrik.  Is it possible to stop/clear a recording with the static API or do you need use the IAnsiConsole in your second example by instantiating a new recorder each time?

Thanks!

**Comment by Sire** [comment]
Thanks Patrik! I will try the Recorder, and if possible build a ToAscii method.

**Comment by codenotworking** [comment]
I'm supplementing something because my needs were the opposite, and if someone else might find this information useful.

I am using `Interprocess Communication (IPC)`, namely [`NamedPipeClientStream`](https://learn.microsoft.com/en-us/dotnet/api/system.io.pipes.namedpipeclientstream) and [`NamedPipeServerStream`](https://learn.microsoft.com/en-us/dotnet/api/system.io.pipes.namedpipeserverstream), in conjunction with [`StreamReader`](https://learn.microsoft.com/en-us/dotnet/api/system.io.streamreader) and [`StreamWriter`](https://learn.microsoft.com/en-us/dotnet/api/system.io.streamwriter), I needed the ANSI colors for a renderable to be intact when sent using IPC.

`AnsiConsole.ExportText()` wasn't keeping the colors.  I had to create a new `IAnsiConsole`:

```cs
// NOTE: You do not need to set any of the properties for AnsiConsoleSettings() for it to work
IAnsiConsole consoleCustom = AnsiConsole.Create( new AnsiConsoleSettings() );

AnsiConsole.Write( panel );

// Writer is a StreamWriter instance sent to either NamedPipeClientStream or NamedPipeServerStream
Writer.WriteLine( consoleCustom.ToAnsi(panel) );
```
